### PR TITLE
libroach: port hash/crc32's ieee implementation to c

### PR DESCRIPTION
**Pull request by benesch**
This is a Friday project to validate a hunch I have that moving all of restore's SST work into C++ will yield a nice performance boost. Getting a fast CRC32 implementation was step 1. Please don't give this a thorough review yet—I haven't had the chance to ask @danhhz if this is even a road we want to seriously explore.

Only the last commit here is out for (proof-of-concept only) review; the first two commits are tracked in #17035 and #17039 respectively.

---

Port Go's highly-optimized AMD64 implementation of CRC32 with the IEEE polynomial to C.

Go 1.9beta2:
```
BenchmarkCRC32/poly=IEEE/size=15/align=0-8         	30000000	        52.0 ns/op	 288.37 MB/s
BenchmarkCRC32/poly=IEEE/size=15/align=1-8         	30000000	        51.9 ns/op	 289.20 MB/s
BenchmarkCRC32/poly=IEEE/size=40/align=0-8         	30000000	        47.5 ns/op	 842.13 MB/s
BenchmarkCRC32/poly=IEEE/size=40/align=1-8         	30000000	        47.5 ns/op	 841.89 MB/s
BenchmarkCRC32/poly=IEEE/size=512/align=0-8        	20000000	        66.3 ns/op	7721.03 MB/s
BenchmarkCRC32/poly=IEEE/size=512/align=1-8        	20000000	        66.7 ns/op	7679.31 MB/s
BenchmarkCRC32/poly=IEEE/size=1kB/align=0-8        	20000000	       111 ns/op	9217.26 MB/s
BenchmarkCRC32/poly=IEEE/size=1kB/align=1-8        	20000000	       115 ns/op	8840.18 MB/s
BenchmarkCRC32/poly=IEEE/size=4kB/align=0-8        	 5000000	       392 ns/op	10443.18 MB/s
BenchmarkCRC32/poly=IEEE/size=4kB/align=1-8        	 5000000	       342 ns/op	11969.82 MB/s
BenchmarkCRC32/poly=IEEE/size=32kB/align=0-8       	  500000	      2574 ns/op	12730.15 MB/s
BenchmarkCRC32/poly=IEEE/size=32kB/align=1-8       	  500000	      2599 ns/op	12603.22 MB/s
```

C:

```
BenchmarkCRC32/poly=IEEE/size=15/align=0-8		 30000000	23.5 ns/op	608.08 MB/s
BenchmarkCRC32/poly=IEEE/size=15/align=1-8		 30000000	23.0 ns/op	621.36 MB/s
BenchmarkCRC32/poly=IEEE/size=40/align=0-8		 30000000	37.4 ns/op	1018.66 MB/s
BenchmarkCRC32/poly=IEEE/size=40/align=1-8		 30000000	38.0 ns/op	1002.82 MB/s
BenchmarkCRC32/poly=IEEE/size=512/align=0-8		 20000000	51.1 ns/op	9548.67 MB/s
BenchmarkCRC32/poly=IEEE/size=512/align=1-8		 20000000	50.7 ns/op	9637.17 MB/s
BenchmarkCRC32/poly=IEEE/size=1kB/align=0-8		 20000000	91.3 ns/op	10698.23 MB/s
BenchmarkCRC32/poly=IEEE/size=1kB/align=1-8		 20000000	91.9 ns/op	10625.31 MB/s
BenchmarkCRC32/poly=IEEE/size=4kB/align=0-8		  5000000	332.9 ns/op	11735.48 MB/s
BenchmarkCRC32/poly=IEEE/size=4kB/align=1-8		  5000000	332.3 ns/op	11754.54 MB/s
BenchmarkCRC32/poly=IEEE/size=32kB/align=0-8	   500000	2516.3 ns/op	12418.92 MB/s
BenchmarkCRC32/poly=IEEE/size=32kB/align=1-8	   500000	2624.0 ns/op	11909.40 MB/s
```

benchstat:

```
name                                 old time/op    new time/op     delta
CRC32/poly=IEEE/size=15/align=0-8      53.0ns ± 1%     23.9ns ± 3%   -54.83%  (p=0.008 n=5+5)
CRC32/poly=IEEE/size=15/align=1-8      53.1ns ± 1%     23.7ns ± 3%   -55.31%  (p=0.008 n=5+5)
CRC32/poly=IEEE/size=40/align=0-8      48.2ns ± 1%     28.7ns ± 2%   -40.42%  (p=0.008 n=5+5)
CRC32/poly=IEEE/size=40/align=1-8      48.5ns ± 2%     28.9ns ± 2%   -40.32%  (p=0.008 n=5+5)
CRC32/poly=IEEE/size=512/align=0-8     65.1ns ± 2%     52.0ns ± 1%   -20.21%  (p=0.008 n=5+5)
CRC32/poly=IEEE/size=512/align=1-8     65.7ns ± 1%     52.6ns ± 2%   -19.91%  (p=0.008 n=5+5)
CRC32/poly=IEEE/size=1kB/align=0-8      106ns ± 2%       94ns ± 1%   -11.40%  (p=0.016 n=5+4)
CRC32/poly=IEEE/size=1kB/align=1-8      106ns ± 1%       97ns ± 8%    -8.96%  (p=0.008 n=5+5)
CRC32/poly=IEEE/size=4kB/align=0-8      357ns ± 9%      341ns ± 2%    -4.49%  (p=0.008 n=5+5)
CRC32/poly=IEEE/size=4kB/align=1-8      347ns ± 1%      338ns ± 2%      ~     (p=0.056 n=5+5)
CRC32/poly=IEEE/size=32kB/align=0-8    2.59µs ± 1%     2.63µs ± 5%      ~     (p=0.690 n=5+5)
CRC32/poly=IEEE/size=32kB/align=1-8    2.64µs ± 1%     2.67µs ±13%      ~     (p=0.500 n=5+5)

name                                 old speed      new speed       delta
CRC32/poly=IEEE/size=15/align=0-8     283MB/s ± 1%    598MB/s ± 3%  +111.25%  (p=0.008 n=5+5)
CRC32/poly=IEEE/size=15/align=1-8     282MB/s ± 1%    603MB/s ± 3%  +113.60%  (p=0.008 n=5+5)
CRC32/poly=IEEE/size=40/align=0-8     829MB/s ± 1%   1328MB/s ± 2%   +60.19%  (p=0.008 n=5+5)
CRC32/poly=IEEE/size=40/align=1-8     826MB/s ± 2%   1320MB/s ± 2%   +59.89%  (p=0.008 n=5+5)
CRC32/poly=IEEE/size=512/align=0-8   7.86GB/s ± 2%   9.40GB/s ± 1%   +19.52%  (p=0.008 n=5+5)
CRC32/poly=IEEE/size=512/align=1-8   7.79GB/s ± 1%   9.28GB/s ± 2%   +19.12%  (p=0.008 n=5+5)
CRC32/poly=IEEE/size=1kB/align=0-8   9.58GB/s ± 1%  10.36GB/s ± 0%    +8.14%  (p=0.016 n=5+4)
CRC32/poly=IEEE/size=1kB/align=1-8   9.60GB/s ± 1%  10.12GB/s ± 8%      ~     (p=0.151 n=5+5)
CRC32/poly=IEEE/size=4kB/align=0-8   11.5GB/s ± 8%   11.5GB/s ± 2%      ~     (p=0.310 n=5+5)
CRC32/poly=IEEE/size=4kB/align=1-8   11.8GB/s ± 1%   11.6GB/s ± 3%      ~     (p=0.310 n=5+5)
CRC32/poly=IEEE/size=32kB/align=0-8  12.6GB/s ± 1%   11.9GB/s ± 5%    -5.76%  (p=0.008 n=5+5)
CRC32/poly=IEEE/size=32kB/align=1-8  12.4GB/s ± 1%   11.8GB/s ±12%    -5.44%  (p=0.008 n=5+5)
```

### CODE REVIEW

**Comment by bdarnell** [comment]
RocksDB already has a crc32 implementation. Could we use that one (looks like we'd need to export the header) instead of introducing a new one?

**Comment by danhhz** [comment]
@bdarnell it's the wrong crc32. They have Castagnoli. If we use this, it would be for verifying our individual kv checksums, which are IEEE

**Comment by danhhz** [comment]
Cool stuff, you'll definitely want someone that speaks c++ to do the review.

I'd be interested to see how this affects the AddSSTable benchmarks. Given this, it should be pretty easy to move the current impl of verifySSTable entirely to c++ and see what happens to BenchmarkAddSSTable and BenchmarkImport/AddSSTable

**Comment by petermattis** [comment]
:lgtm: modulo using curly braces always.

@raduberinde Can you take a look at the crc stuff given your familiarity with the Go crc code?

---

Review status: 0 of 51 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[c-deps/libroach/LICENSE.go, line 1 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/17040#-KpGiAabPrhs7jAwVBUj:-KpGiAabPrhs7jAwVBUk:bewioin) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb11d984700e79ebb9edfb1b35b61bbf7e7ffed2/c-deps/libroach/LICENSE.go#L1)):*
> ```Go
> Copyright (c) 2009 The Go Authors. All rights reserved.
> ```

Giving this file the `.go` suffix is a bit confusing since that makes it appear to be a Go source file. Perhaps `LICENSE.golang`. 

---

*[c-deps/libroach/testutil.c, line 24 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/17040#-KpGiO6Mp7HjPtCE1SGR:-KpGiO6Mp7HjPtCE1SGS:bacw25m) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb11d984700e79ebb9edfb1b35b61bbf7e7ffed2/c-deps/libroach/testutil.c#L24)):*
> ```C
> make_bytes(size_t len) {
> 	long* buf = malloc(len);
> 	if (buf == NULL)
> ```

Almost all C/C++ style guidelines recommend using curly braces even when they aren't necessary. For example, `if (buf == NULL) { err(...); }`

Also, you seem to be using tabs for indentation in this file while other files use spaces. I'd stick with spaces.

---

*[c-deps/libroach/testutil.h, line 17 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/17040#-KpGj0XzyoRGgP8udweb:-KpGj0XzyoRGgP8udwec:ba7t5w) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb11d984700e79ebb9edfb1b35b61bbf7e7ffed2/c-deps/libroach/testutil.h#L17)):*
> ```objc
> // Author: Nikhil Benesch (nikhil.benesch@gmail.com)
> 
> #define ARRAY_SIZE(a) (sizeof(a) / sizeof(a[0]))
> ```

You only seem to use this in `crc32_bench.c`. Move the definition there?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17040#-:-KpGjIRSM2eYjU2-AaN-:bpv61da)*

**Comment by tamird** [comment]
I'd like to review this - could you ping when it's rebased on top of #17035?

---

Review status: 0 of 51 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17040#-:-KpI5BzrsDLbqmCGF0LF:b-qndbe8)*

**Comment by benesch** [comment]
Ping!

Rebuilding the builder now.

---

Review status: 0 of 15 files reviewed at latest revision, 3 unresolved discussions.

---

*[c-deps/libroach/testutil.c, line 24 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/17040#-KpGiO6Mp7HjPtCE1SGR:-KpI90zGyWe9mnKWuayJ:b-glx2gt) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb11d984700e79ebb9edfb1b35b61bbf7e7ffed2/c-deps/libroach/testutil.c#L24)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Almost all C/C++ style guidelines recommend using curly braces even when they aren't necessary. For example, `if (buf == NULL) { err(...); }`

Also, you seem to be using tabs for indentation in this file while other files use spaces. I'd stick with spaces.
</blockquote></details>

Yeah, it's a bad habit I've picked up, and there's no go fmt to tell me I can't. Done and done.

---

*[c-deps/libroach/testutil.h, line 17 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/17040#-KpGj0XzyoRGgP8udweb:-KpI97v0pTyoW_POsigk:be74cn7) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb11d984700e79ebb9edfb1b35b61bbf7e7ffed2/c-deps/libroach/testutil.h#L17)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

You only seem to use this in `crc32_bench.c`. Move the definition there?
</blockquote></details>

It's used in crc32_test.c as well.

---

*[c-deps/libroach/LICENSE.go, line 1 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/17040#-KpGiAabPrhs7jAwVBUj:-KpI9VBz4n42x2XcsmSU:bombmv2) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb11d984700e79ebb9edfb1b35b61bbf7e7ffed2/c-deps/libroach/LICENSE.go#L1)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Giving this file the `.go` suffix is a bit confusing since that makes it appear to be a Go source file. Perhaps `LICENSE.golang`. 
</blockquote></details>

Good call. Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17040#-:-KpI9uAOsZZEZhJVqNYq:b-o3baqt)*

**Comment by RaduBerinde** [comment]
Have you considered lifting the C implementations from the linux kernel instead? This kind of code is usually highly optimized there (in some cases, Intel contributes directly). I'm not sure if the GPL license is an issue.

https://github.com/torvalds/linux/blob/master/arch/x86/crypto/crc32-pclmul_asm.S

**Comment by benesch** [comment]
My understanding is we can't include or link any GPL-licensed code or else
we become a derived work and must license Cockroach under the GPL, too. (In
fact, I specifically *didn't* look at the Linux implementation.)

File: c-deps/libroach/crc32_amd64.S [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+ * operate on two double quadword (64-bit) operands at a time, and is described
+ * in detail in an Intel whitepaper, "Fast CRC Computation for Generic
+ * Polynomials Using PCLMULQDQ Instruction" [1].
+ *
+ * [0]: https://github.com/golang/go/blob/f3e6216450866f761cc22c042798c88062319867/src/hash/crc32/crc32_amd64.s
+ * [1]: https://www.intel.com/content/dam/www/public/us/en/documents/white-papers/fast-crc-computation-generic-polynomials-pclmulqdq-paper.pdf
+ */
+
+/*
+ * These constants are lifted from the Linux kernel, since they avoid the costly
```
Comment (RaduBerinde) [inline c-deps/libroach/crc32_amd64.S]: Looks like the go folks already looked into that :)

**Comment by RaduBerinde** [comment]
LGTM. Haven't looked at the assembler in detail, but the tests are great (especially the cross-checks).

Too bad we're not using CRC32-C which is faster than CRC32-IEEE.

File: build/teamcity-check.sh [hunk 2]
```diff
@@ -18,7 +18,8 @@ build/builder.sh make lint 2>&1 | tee artifacts/lint.log | go-test-teamcity
 build/builder.sh make generate
 build/builder.sh /bin/bash -c '! git status --porcelain | read || (git status; git diff -a 1>&2; exit 1)'
 
-# Run the UI tests. This logically belongs in teamcity-test.sh, but we do it
-# here to minimize total build time since the rest of this script completes
-# faster than the non-UI tests.
+# Run the UI tests and libroach tests. These logically belong in
+# teamcity-test.sh, but we do it here to minimize total build time since the
+# rest of this script completes much faster than teamcity-test.sh.
```
Comment (dt) [inline build/teamcity-check.sh]: off-topic: this is getting less true, and indeed i'd argue it's about time to make a `teamcity-ui.sh` or `teamcity-generate.sh`

**Comment by benesch** [comment]
We're going to get rid of the CRC32C checksums entirely, so no need for this now!

**Pull request closed without merging**